File: .e2e_v/concurrency_SyncCounter/SyncCounter.go
package main

import (
	fmt "fmt"
	stdjava "github.com/NickyBoy89/java2go/stdjava"
	os "os"
)

type SyncCounter struct {
	count int32
}

func (sr *SyncCounter) __java2goInitFields() {
	sr.count = 0
}
func NewSyncCounter() *SyncCounter {
	sr := new(SyncCounter)
	sr.__java2goInitFields()
	return sr
}
func (sr *SyncCounter) increment() {
	__java2goMethodMonitor := stdjava.MonitorEnter(sr)
	defer stdjava.MonitorExit(__java2goMethodMonitor)
	sr.count++
}
func (sr *SyncCounter) get() int32 {
	return sr.count
}
// throws InterruptedException
func Main() {
	args := os.Args
	_ = args
	counter := NewSyncCounter()
	threadCount := int32(4)
	perThread := int32(1000)
	threads := make([]*stdjava.Thread, threadCount)
	for t := int32(0); t < threadCount; t++ {
		threads[int(t)] = stdjava.NewThread(&SyncCounterAnon1{perThread: perThread, counter: counter})
	}
	for t := int32(0); t < threadCount; t++ {
		threads[int(t)].Start()
	}
	for t := int32(0); t < threadCount; t++ {
		threads[int(t)].Join()
	}
	fmt.Println(counter.get())
}

type SyncCounterAnon1 struct {
	perThread	int32
	counter		*SyncCounter
}

func (s1 *SyncCounterAnon1) Run() {
	for i := int32(0); i < s1.perThread; i++ {
		s1.counter.increment()
	}
}

File: testfiles/e2e/concurrency/SyncCounter.java
public class SyncCounter {
    private int count = 0;

    synchronized void increment() {
        count++;
    }

    int get() {
        return count;
    }

    public static void main(String[] args) throws InterruptedException {
        SyncCounter counter = new SyncCounter();
        int threadCount = 4;
        int perThread = 1000;

        Thread[] threads = new Thread[threadCount];
        for (int t = 0; t < threadCount; t++) {
            threads[t] = new Thread(new Runnable() {
                public void run() {
                    for (int i = 0; i < perThread; i++) {
                        counter.increment();
                    }
                }
            });
        }

        for (int t = 0; t < threadCount; t++) {
            threads[t].start();
        }
        for (int t = 0; t < threadCount; t++) {
            threads[t].join();
        }

        System.out.println(counter.get());
    }
}
